File: rest-service/src/main/java/mb/lib/query/db/SelectUserTargetLinks.java
package mb.lib.query.db;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.vulpine.lib.query.util.basic.BasicPreparedReadQuery;
import mb.lib.model.HashID;
import mb.lib.query.model.BlastTargetLink;

public class SelectUserTargetLinks
{
  private static final String Query = """
    SELECT
      a.*
    FROM
      userlogins5.multiblast_job_to_targets a
      INNER JOIN userlogins5.multiblast_users mu
        ON a.job_digest = mu.job_digest
    WHERE
      mu.user_id = ?
    """;

  private final Connection con;
  private final long userID;

  public SelectUserTargetLinks(Connection con, long userID) {
    this.con    = con;
    this.userID = userID;
  }

  public Map<HashID, List<BlastTargetLink>> run() throws Exception {
    return new BasicPreparedReadQuery<>(Query, con, this::parse, this::prep).execute().getValue();
  }

  private Map<HashID, List<BlastTargetLink>> parse(ResultSet rs) throws Exception {
    var out = new HashMap<HashID, List<BlastTargetLink>>();

    while (rs.next()) {
      var tmp = Util.parseTargetLink(rs);

      out.computeIfAbsent(tmp.getJobID(), k -> new ArrayList<>())
        .add(tmp);
    }

    return out;
  }

  private void prep(PreparedStatement ps) throws Exception {
    ps.setLong(1, userID);
  }
}

File: rest-service/src/main/kotlin/mb/lib/query/db/SelectUserTargetLinks.kt
package mb.lib.query.db

import io.vulpine.lib.query.util.basic.BasicPreparedReadQuery
import mb.lib.query.model.BlastTargetLink
import org.veupathdb.lib.hash_id.HashID
import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.ResultSet

data class SelectUserTargetLinks(
  private val con: Connection,
  private val userID: Long,
) {
  companion object {
    // language=postgresql
    private const val Query = """
    SELECT
      a.*
    FROM
      multiblast.multiblast_job_to_targets a
      INNER JOIN multiblast.multiblast_users mu
        ON a.job_digest = mu.job_digest
    WHERE
      mu.user_id = ?
    """
  }

  fun run(): Map<HashID, List<BlastTargetLink>> {
    return BasicPreparedReadQuery(Query, con, this::parse, this::prep).execute().getValue()
  }

  private fun parse(rs: ResultSet): Map<HashID, List<BlastTargetLink>> {
    val out = HashMap<HashID, List<BlastTargetLink>>()

    while (rs.next()) {
      val tmp = parseTargetLink(rs)

      (out.computeIfAbsent(tmp.jobID!!) { ArrayList() } as ArrayList)
        .add(tmp)
    }

    return out
  }

  private fun prep(ps: PreparedStatement) {
    ps.setLong(1, userID)
  }
}
